cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -214,6 +214,62 @@ struct SubtractChecked {
   }
 };
 
+template <bool is_32bit, int64_t multiple>
+struct SubtractTimeDuration {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_t<is_32bit, T> Call(KernelContext*, Arg0 left, Arg1 right,
+                                       Status* st) {
+    T result = arrow::internal::SafeSignedSubtract(left, static_cast<T>(right));
+    if (result < 0) {
+      *st = Status::Invalid(result, " is not within the acceptable range of ", "[0, ",
+                            multiple, ") s");
+    }
```

Comment:
Since we could subtract a negative duration, we should also check if the output is larger than the maximum acceptable value.

Reply:
Alternatively - maybe this arithmetic should wrap around? Does that make any sense?

Reply:
Huh. I wonder how expected this would be. But you're proposing something like this: 
```
if (result < 0) {
  result = result % multiple;
}
```
Correct?

Reply:
Right. Though, I'm not sure if that really makes any sense to have.

Reply:
It feels semantically correct but I'm going to check what Pandas and others do to see what conventions are.

Reply:
I poked around stdlib datetime but it doesn't seem to implement this. However, I still think we need to check whether the result is in range on both ends.

Reply:
I found this reference: https://bugs.python.org/issue1487389#msg54803

I think raising an error is OK. If we need wrap-around, we can add that as a separate kernel perhaps.

Reply:
We should still raise an error if the result is >24:00 though.

Reply:
Thanks for looking into this! Pandas appears delegate time to Python and rather work with timedelta which is like our duration.
I made this raise on time < 0 and time >= 24:00. Lets revisit if/when desired.